udtube/data/collators.py

```diff
@@ -10,47 +12,100 @@
 from . import batches, datasets
 
 
-class Tokenizer:
-    """Helper for tokenization.
+class Error(Exception):
+    pass
 
-    Args:
-        tokenizer: transformer autotokenizer.
-    """
 
-    tokenizer: transformers.AutoTokenizer
-
-    def __init__(self, tokenizer):
-        self.tokenizer = tokenizer
-
-    def __call__(self, tokens: List[List[str]]) -> transformers.BatchEncoding:
-        """Runs tokenizer.
+@dataclasses.dataclass
+class Collator:
+    """Collator for CoNLL-U data."""
 
-        Args:
-            tokens: a list of sentences, each a list of tokens.
+    tokenizer: transformers.AutoTokenizer
 
-        Returns:
-            The BatchEncoding object.
-        """
-        return self.tokenizer(
-            tokens,
+    def __call__(self, itemlist: List[datasets.Item]) -> batches.Batch:
+        # Runs the tokenizer.
+        tokenized = self.tokenizer(
+            [item.get_tokens() for item in itemlist],
             padding="longest",
+            truncation=False,
             return_tensors="pt",
             is_split_into_words=True,
             add_special_tokens=False,
         )
+        actual_length = tokenized.input_ids.size(1)
+        max_length = self.tokenizer.model_max_length
+        if actual_length > max_length:
+            # By construction, a sequence is too long if the first element
+            # beyond the encoder's max length is not padding.
+            keep_items = tokenized.input_ids[:, max_length] == special.PAD_IDX
+            # Ideally we'd just shorten the tag tensors, but mapping subword
+            # tokens to tags is sufficiently complex that we don't know at
+            # this stage how many tags to keep or get rid of. Therefore we
+            # just discard these sentences.
+            logging.warning(
+                "Discarding %d sequence(s) exceeding the encoder's "
+                "maximum length (%d)",
+                torch.sum(~keep_items),
+                max_length,
+            )
+            itemlist = self._keep_list(itemlist, keep_items)
+            # It is possible but very unlikely that we will now have a
+            # completely empty batch. Since this will result in numerous errors
+            # downstream no matter what we do, we raise an error.
+            if not itemlist:
+                raise Error("Entire batch was truncated")
```

Comment:
Very opinionated/preference based, but I feel this isn't very refective of what happens - 

`"Every sequence in the batch was discarded due to exceeding the model's max input length "`